### PR TITLE
Feature: Add ability to mutate navigational breadcrumbs

### PR DESCRIPTION
**Pull request by ueman**
## :scroll: Description
This adds the option to mutate `RouteSettings` via callback in the Navigation Observer and it also adds the ability to add some context to each navigational breadcrumb.

## :bulb: Motivation and Context
Flutter allowes you to nest `Navigator`s. In order to differentiate breadcrumbs, I added the ability to do

```dart
SentryNavigatorObserver(
    additionalInfoProvider: (from, to) {
        return <String, dynamic>{'navigator': 'root-navigator'};
    },
);
``` 

Sometimes you don't want specific routes to be included, for example if they come from third party dependencies or you want your routes to be renamed to match analytic events. So you're now able to:

```dart
SentryNavigatorObserver(
    routeNameExtractor: (routeSettings) {
        return routeSettings.copyWith(name: 'different name');
        // or reset name to `null`:
       return RouteSettings(arguments: routeSettings.arguments);
    },
);
``` 

## :green_heart: How did you test it?
New tests

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [ ] I updated the docs if needed
- [ ] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: flutter/lib/src/navigation/sentry_navigator_observer.dart [hunk 1]
```diff
@@ -8,6 +8,15 @@ import '../../sentry_flutter.dart';
 /// See https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/
 const _navigationKey = 'navigation';
 
+typedef RouteNameExtractor = RouteSettings Function(RouteSettings settings);
+
+typedef AdditionalInfoProvider = Map<String, dynamic>? Function(
```
Comment (ueman) [inline flutter/lib/src/navigation/sentry_navigator_observer.dart]: I'm not quite sure about these names, any suggestions?

Comment (brustolin) [reply flutter/lib/src/navigation/sentry_navigator_observer.dart]: I would change it from Extractor and Provider to Callback.
'Extractor' is very odd, and Provider is usually associated with other things in Flutter.

Comment (ueman) [reply flutter/lib/src/navigation/sentry_navigator_observer.dart]: Sounds good

**Comment by marandaneto** [comment]
@ueman before going ahead, wouldn't be possible to do this using the before breadcrumb callback? https://docs.sentry.io/platforms/flutter/enriching-events/breadcrumbs/#customize-breadcrumbs

**Comment by ueman** [comment]
> @ueman before going ahead, wouldn't be possible to do this using the before breadcrumb callback? https://docs.sentry.io/platforms/flutter/enriching-events/breadcrumbs/#customize-breadcrumbs

Unfortunately not, at that point you're not able to know which `Navigator` the breadcrumb comes from.


Some  more  information on why this is useful:
You're able to nest `Navigator`s. For example, you can have one navigator for the app, which has a Tab Bar UI. Each tab then again has its own navigator. Each Navigator could theoretically have the same routes as the other ones. 

So this makes it useful to add some information at the `NavigationObserver` level because you're now able to associate information about which navigator is used.

I hope that clarifies a little bit.

**Comment by marandaneto** [comment]
@ueman looks legit then, a last question, what if we pass the `RouteSettings` to the `hint` param from `addBreadcrumb`? it's meant for this type of usage, we set automatically everything that would make sense, but the user can still mutate the breadcrumb using the before breadcrumb callback, the `hint` is actually where the data was set from.

**Comment by ueman** [comment]
> @ueman looks legit then, a last question, what if we pass the `RouteSettings` to the `hint` param from `addBreadcrumb`? it's meant for this type of usage, we set automatically everything that would make sense, but the user can still mutate the breadcrumb using the before breadcrumb callback, the `hint` is actually where the data was set from.

RouteSettings still don't specify which Navigator is its origin :(

**Review by marandaneto — APPROVED** [review]

**Comment by marandaneto** [comment]
thanks for doing this @ueman !

**Pull request merged**